clients/R/R/tinystan.R

```diff
@@ -1,51 +1,133 @@
-    private$lib_name <- tools::file_path_sans_ext(basename(lib))
-    if (warn && is.loaded("tinystan_create_model_R", PACKAGE = private$lib_name)) {
+    lib <- tools::file_path_as_absolute(lib)
+    lib_name <- tools::file_path_sans_ext(basename(lib))
+    if (warn && is.loaded("tinystan_create_model_R", PACKAGE = lib_name)) {
         warning(paste0("Loading a shared object '", lib, "' which is already loaded.\n",
             "If the file has changed since the last time it was loaded, this load may not update the library!"))
     }
+    dyn.load(lib, PACKAGE = lib_name)
+    sep <- .C("tinystan_separator_char_R", sep = raw(1), PACKAGE = lib_name)$sep
+    sep <- rawToChar(sep)
+    ret <- list(lib = lib, lib_name = lib_name, sep = sep, code = code, built_with_so = built_with_so)
+    class(ret) <- c("tinystan_model", class(ret))
+    return(ret)
+}
 
-    dyn.load(private$lib, PACKAGE = private$lib_name)
-
-    sep <- .C("tinystan_separator_char_R", sep = raw(1), PACKAGE = private$lib_name)$sep
-    private$sep <- rawToChar(sep)
+#' @export
+print.tinystan_model <- function(mod, ...) {
+    cat(mod$code, ...)
+    if (mod$built_with_so) {
+        cat("Library: ", mod$lib, "\n")
+    }
+}
 
-}, api_version = function() {
+#'@export
+api_version = function(stan_model) {
     .C("tinystan_api_version", major = integer(1), minor = integer(1), patch = integer(1),
-        PACKAGE = private$lib_name)
-}, sample = function(data = "", num_chains = 4, inits = NULL, seed = NULL, id = 1,
-    init_radius = 2, num_warmup = 1000, num_samples = 1000, metric = HMCMetric$DIAGONAL,
-    init_inv_metric = NULL, save_metric = FALSE, adapt = TRUE, delta = 0.8, gamma = 0.05,
-    kappa = 0.75, t0 = 10, init_buffer = 75, term_buffer = 50, window = 25, save_warmup = FALSE,
-    stepsize = 1, stepsize_jitter = 0, max_depth = 10, refresh = 0, num_threads = -1) {
+        PACKAGE = stan_model$lib_name)
+}
+
+#' @noRd
+with_model = function(private, data, seed, block) {
+    ffi_ret <- .C("tinystan_create_model_R", model = raw(8), as.character(data),
+        as.integer(seed), err = raw(8), NAOK = TRUE, PACKAGE = private$lib_name)
+    handle_error(all(ffi_ret$model == 0), private$lib_name, ffi_ret$err)
+    tryCatch({
+        # this is the equivalent of base R's `with` function
+        eval(substitute(block), ffi_ret, enclos = parent.frame())
+    }, finally = {
+        .C("tinystan_destroy_model_R", as.raw(ffi_ret$model), PACKAGE = private$lib_name)
+    })
+}
+
+#' @noRd
+get_parameter_names <- function(...) {
+    UseMethod("get_parameter_names")
+}
+
+#' @noRd
+get_parameter_names.tinystan_model = function(private, model) {
+    param_names_raw <- .C("tinystan_model_param_names_R", as.raw(model), names = as.character(""),
+        PACKAGE = private$lib_name)$names
+    if (param_names_raw == "") {
+        return(c())
+    }
+    strsplit(param_names_raw, ",")[[1]]
+
+}
+
+#' @noRd
+get_free_params <- function(...) {
+    UseMethod("get_free_params")
+}
```

Comment:
Do these two need to be S3 methods? Does it matter?

Reply:
My thought is that it's nicer in S3 to define a base function as a generic and then overload it for your class. Since functions from libraries are all loaded into the global namespace if another package has `get_free_params` and we both define it as a generic with overloads everyones code still works as intended instead of overwriting each other

Reply:
Makes sense. 

Aside: These end up in the global namespace even though they aren't exported? Joy